components/raftstore/src/store/worker/snap_gen.rs

```diff
@@ -0,0 +1,278 @@
+}
+
+impl<EK, R, T> Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK>,
+    T: PdClient + 'static,
+{
+    pub fn new(
+        engine: EK,
+        mgr: SnapManager,
+        router: R,
+        pd_client: Option<Arc<T>>,
+        pool: FuturePool,
+    ) -> Runner<EK, R, T> {
+        Runner {
+            tiflash_stores: HashMap::default(),
+            engine,
+            mgr,
+            router,
+            pd_client,
+            pool,
+            semaphore: Arc::new(Semaphore::new(SNAP_GENERATOR_MAX_POOL_SIZE)),
+        }
+    }
+}
+
+impl<EK, R, T> Runnable for Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK> + Send + Clone + 'static,
+    T: PdClient,
+{
+    type Task = Task<EK::Snapshot>;
+
+    fn run(&mut self, task: Task<EK::Snapshot>) {
+        match task {
+            Task::Gen {
+                region_id,
+                last_applied_term,
+                last_applied_state,
+                kv_snap,
+                canceled,
+                notifier,
+                for_balance,
+                to_store_id,
+            } => {
+                // It is safe for now to handle generating and applying snapshot concurrently,
+                // but it may not when merge is implemented.
+                let mut allow_multi_files_snapshot = false;
+                // if to_store_id is 0, it means the to_store_id cannot be found
+                if to_store_id != 0 {
+                    if let Some(is_tiflash) = self.tiflash_stores.get(&to_store_id) {
+                        allow_multi_files_snapshot = !is_tiflash;
+                    } else {
+                        let is_tiflash = self.pd_client.as_ref().map_or(false, |pd_client| {
+                            if let Ok(s) = pd_client.get_store(to_store_id) {
+                                return s.get_labels().iter().any(|label| {
+                                    label.get_key().to_lowercase() == ENGINE
+                                        && label.get_value().to_lowercase() == TIFLASH
+                                });
+                            }
+                            true
+                        });
+                        self.tiflash_stores.insert(to_store_id, is_tiflash);
+                        allow_multi_files_snapshot = !is_tiflash;
+                    }
+                }
+                SNAP_COUNTER.generate.all.inc();
+                let ctx = SnapGenContext {
+                    engine: self.engine.clone(),
+                    mgr: self.mgr.clone(),
+                    router: self.router.clone(),
+                    start: UnixSecs::now(),
+                };
+
+                let scheduled_time = Instant::now_coarse();
+                let semaphore = self.semaphore.clone();
+                self.pool.spawn(async move {
+                    // Limit the concurrency of the snapshot generation tasks.
+                    let _permit = semaphore.acquire().await;
```

Comment:
add a pending task metrics

Reply:
There's an existing `tikv_futurepool_pending_task_total` metric that tracks the number of tasks in a future pool. I think that should be sufficient for identifying task accumulation? 

There's also a grafana panel for it already:
![O3klM6mXSg](https://github.com/user-attachments/assets/f3b4e03e-4ad2-4a34-827e-b3263a032578)